### DISCUSSION TITLE
Who will be the GitHub actor when a workflow runs on a schedule?

### DISCUSSION THREAD

**Discussion by localheinz** [community/community#25067] [Actions]
<p>The documentation at <a href="https://help.github.com/en/actions/automating-your-workflow-with-github-actions/contexts-and-expression-syntax-for-github-actions#github-context" rel="nofollow noopener">https://help.github.com/en/actions/automating-your-workflow-with-github-actions/contexts-and-expression-syntax-for-github-actions#github-context</a> states that github.actor resolves to <em>The login of the user that initiated the workflow run</em>.</p>
<p>Now, who would that be when the workflow runs on a schedule?</p>
<p>I noticed in <a href="https://github.com/ergebnis/composer-normalize/runs/388922221" rel="nofollow noopener">https://github.com/ergebnis/composer-normalize/runs/388922221</a> (a scheduled run) that the actor resolves to localheinz (that’s me).</p>

**Comment by weide-zhou** [comment, marked as answer]
<p>The latest user who create the schedule trigger will be the actor for the schedule workflow.</p>
<p>As mentioned, you can export ‘github.actor’ in context for check.</p>
<p>For instance, if user A create the schedule trigger, workflow is triggered with actor ‘user A’, then user B modified the schedule setting, workflow is triggered with actor ‘user B’.</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/f/fbbc1ff46d030e5a4e29a33fbdaae1ccb9896510.png" data-download-href="/uploads/short-url/zUWTQ80BlkBNB9rf3z3EvvLk3jW.png?dl=1" title="scheduleactor.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/52772761/181088644-09833582-65bf-4426-957f-883d5423dd8b.png" alt="scheduleactor.png" data-base62-sha1="zUWTQ80BlkBNB9rf3z3EvvLk3jW" width="467" height="500" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/f/fbbc1ff46d030e5a4e29a33fbdaae1ccb9896510_2_467x500.png, https://aws1.discourse-cdn.com/github/optimized/2X/f/fbbc1ff46d030e5a4e29a33fbdaae1ccb9896510_2_700x750.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/f/fbbc1ff46d030e5a4e29a33fbdaae1ccb9896510.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/f/fbbc1ff46d030e5a4e29a33fbdaae1ccb9896510_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">scheduleactor.png</span><span class="informations">765×818 37.3 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Reply by henninge** [reply]
So "git blame" for our workflow currently reads

```
aaaaaaaaa (User A 2022-01-13 20:22:28 +0100  21)   schedule:
bbbbbbbbb (User B 2022-05-04 17:36:23 +0200  22)     - cron: "45 */4 * * *"
``` 

and User A shows up as the actor in the UI, like this:

```
Triggered via schedule 3 hours ago
User A  bbbbbbbbb
```

Or is the user in the UI not the actor? It seems that whoever added the `schedule:` in the first place gets listed.

**Reply by sxlijin** [reply]
Just ran a bunch of tests of this. The user in the blame is not guaranteed to be the same as the actor; it just usually happens to be the same by coincidence.

The real answer is this: the identity of the `actor` is the identity used for the `push` operation that last updated the workflow.

In other words, it has absolutely nothing to do with the `author` or `committer` of the latest commit to update the workflow. This makes sense, in a confusing way: from GH's perspective, when a workflow is updated, the only actual "identity" it has is the authenticated identity backing the authorization claim on the `push` operation.

In fact, GitHub fundamentally cannot use the `author` or `committer` to infer a principal (identity), because commits generally contain no proof about the identity of the author or committer (signed commits being the exception. technically, but even then, you can't use a signed commit to resolve a principal; you use the public key associated with a principal to verify the signature). When the GitHub UI links to an account claiming "USER committed on Oct 3" or lists USER in the blame for a file, what GitHub is really doing is making a best-effort association from the `author.email` of the commit in question to a GitHub user. This is a purely best-effort operation with no authn/z implications. which is why [GitHub Security explicitly considers email address impersonation to be a non-issue](https://bounty.github.com/ineligible.html#impersonating_a_user_through_git_email_address).

By contrast, workflows are - I assume - entities that exist somewhere in a GitHub database which can only be updated by authorized operations. Using the principal of the authorization claim on the most recent update as the `actor` for a scheduled workflow is pretty logical to me, if non-obvious.

Here's a test I ran of this, where the commit that updated the workflow file was created with the email address of `sxlijin1`:

![image](https://github.com/community/community/assets/512410/f122d557-81b3-4d81-90b4-631ebf49ca31) ![image](https://github.com/community/community/assets/512410/5080498c-0156-4ea1-8e93-6b9c592555c8)

**Comment by bparthas21** [comment]
I think this answer is wrong. In my experience, it has *nothing* to do with who edited the `cron` schedule. All it matters is who edited the file **last** immaterial of what line was changed in the workflow file. @localheinz @weide-zhou

**Comment by dhalbert** [comment]
I asked GitHub support about this and got the following answer:

> Indirectly mentioned in this [documentation snippet](https://docs.github.com/en/actions/monitoring-and-troubleshooting-workflows/notifications-for-workflow-runs), the actor tied to a workflow run triggered on schedule is the user who last edited the cron syntax on the workflow file, or re-enabled the workflow after it has been disabled.
> 
>     Notifications for scheduled workflows are sent to the user who initially created the workflow. If a different user updates the cron syntax in the workflow file, subsequent notifications will be sent to that user instead. If a scheduled workflow is disabled and then re-enabled, notifications will be sent to the user who re-enabled the workflow rather than the user who last modified the cron syntax.
> 
> In this case, you can have the `@[desired new user]` [disable and re-enable the workflow](https://docs.github.com/en/actions/using-workflows/disabling-and-enabling-a-workflow) so that they become the actor for further workflow runs.
> 

I had first tried to change the actor by merging a PR by the desired new user to the cron job .yml action. That did not change the actor. What did work was to log in as the new user and then disable and re-enable the workflow, as linked to above.

**Comment by bejgli** [comment]
This is very inconvenient when it comes to sending workflow run notifications to a specific user. Are there any plans for an explicit "actor" option for scheduled workflows?

**Reply by younsl** [reply]
I think so. It’s uncomfortable because it’s not intuitive.
